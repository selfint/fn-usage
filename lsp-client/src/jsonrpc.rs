use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Request<Params> {
    pub jsonrpc: String,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Params>,
    pub id: i64,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Notification<Params> {
    pub jsonrpc: String,
    pub method: String,
    pub params: Option<Params>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Response<T, E> {
    pub jsonrpc: String,
    #[serde(flatten)]
    pub result: JsonRpcResult<T, E>,
    pub id: Option<i64>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "lowercase")]
pub enum JsonRpcResult<T, E> {
    Result(T),
    Error {
        code: i64,
        message: String,
        data: Option<E>,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_request_serialization() {
        insta::assert_compact_json_snapshot!(
            Request {
                jsonrpc: "2.0".to_string(),
                method: "method".to_string(),
                params: Some(vec![42, 23]),
                id: 1,
            },
            @r###"{"jsonrpc": "2.0", "method": "method", "params": [42, 23], "id": 1}"###
        );

        // a but awkward, as Some(()) should probably just no be serialized
        // implementing this requires a lot of ugly code, so won't be implemented
        // unless it is a problem
        insta::assert_compact_json_snapshot!(
            Request {
                jsonrpc: "2.0".to_string(),
                method: "method".to_string(),
                params: Some(()),
                id: 1,
            },
            @r###"{"jsonrpc": "2.0", "method": "method", "params": null, "id": 1}"###
        );
    }

    #[test]
    fn test_request_deserialization() {
        insta::assert_debug_snapshot!(
            serde_json::from_str::<Request<Vec<i32>>>(r#"{"jsonrpc": "2.0", "method": "method", "params": [42, 23], "id": 1}"#),
            @r###"
        Ok(
            Request {
                jsonrpc: "2.0",
                method: "method",
                params: Some(
                    [
                        42,
                        23,
                    ],
                ),
                id: 1,
            },
        )
        "###
        );
    }

    #[test]
    fn test_notification_serialization() {
        insta::assert_compact_json_snapshot!(
            Notification {
                jsonrpc: "2.0".to_string(),
                method: "method".to_string(),
                params: Some(vec![42, 23]),
            },
            @r###"{"jsonrpc": "2.0", "method": "method", "params": [42, 23]}"###
        );
    }

    #[test]
    fn test_notification_deserialization() {
        insta::assert_debug_snapshot!(
            serde_json::from_str::<Notification<Vec<i32>>>(r#"{"jsonrpc": "2.0", "method": "method", "params": [42, 23]}"#),
            @r###"
        Ok(
            Notification {
                jsonrpc: "2.0",
                method: "method",
                params: Some(
                    [
                        42,
                        23,
                    ],
                ),
            },
        )
        "###
        );
    }

    #[test]
    fn test_response_serialization() {
        insta::assert_compact_json_snapshot!(
            Response {
                jsonrpc: "2.0".to_string(),
                result: JsonRpcResult::<_, ()>::Result(19),
                id: Some(1),
            },
            @r###"{"jsonrpc": "2.0", "result": 19, "id": 1}"###
        );

        insta::assert_compact_json_snapshot!(
            Response {
                jsonrpc: "2.0".to_string(),
                result: JsonRpcResult::Error::<(), _>  {
                    code: -32601,
                    message: "Method not found".to_string(),
                    data: Some(vec!["Some", "data"])
                },
                id: None,
            },
            @r###"{"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found", "data": ["Some", "data"]}, "id": null}"###
        );
    }

    #[test]
    fn test_response_deserialization() {
        insta::assert_debug_snapshot!(
            serde_json::from_str::<Response<i32, ()>>(r#"{"jsonrpc": "2.0", "result": 19, "id": 1}"#),
            @r###"
        Ok(
            Response {
                jsonrpc: "2.0",
                result: Result(
                    19,
                ),
                id: Some(
                    1,
                ),
            },
        )
        "###
        );

        insta::assert_debug_snapshot!(
            serde_json::from_str::<Response<(), Vec<String>>>(r#"{"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found", "data": ["Some", "data"]}, "id": null}"#),
            @r###"
        Ok(
            Response {
                jsonrpc: "2.0",
                result: Error {
                    code: -32601,
                    message: "Method not found",
                    data: Some(
                        [
                            "Some",
                            "data",
                        ],
                    ),
                },
                id: None,
            },
        )
        "###
        );
    }
}
