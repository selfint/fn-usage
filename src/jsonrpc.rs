use serde::{Deserialize, Serialize, de::DeserializeOwned};
use serde_json::Value;

#[derive(Serialize, Deserialize, Debug)]
pub struct Request<Params> {
    pub jsonrpc: String,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Params>,
    pub id: i64,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Notification<Params> {
    pub jsonrpc: String,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Params>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Response<T: Serialize + DeserializeOwned> {
    pub jsonrpc: String,
    #[serde(flatten)]
    #[serde(with = "JsonRpcResult")]
    pub result: Result<T, Error>,
    pub id: Option<i64>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Error {
    code: i64,
    message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    data: Option<Value>,
}

type Remote<T> = Result<T, Error>;

#[derive(Serialize, Deserialize, Debug)]
#[serde(remote = "Remote")]
enum JsonRpcResult<T> {
    #[serde(rename = "result")]
    Ok(T),
    #[serde(rename = "error")]
    Err(Error),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Error {}: {}", self.code, self.message)
    }
}

impl std::error::Error for Error {}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn test_request_serialization() {
        insta::assert_compact_json_snapshot!(
            Request {
                jsonrpc: "2.0".to_string(),
                method: "subtract".to_string(),
                params: Some(vec![42, 23]),
                id: 1,
            },
            @r#"{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}"#
        );

        // a but awkward, as Some(()) should probably just no be serialized
        // implementing this requires a lot of ugly code, so won't be implemented
        // unless it is a problem
        insta::assert_compact_json_snapshot!(
            Request {
                jsonrpc: "2.0".to_string(),
                method: "method".to_string(),
                params: Some(()),
                id: 1,
            },
            @r###"{"jsonrpc": "2.0", "method": "method", "params": null, "id": 1}"###
        );

        insta::assert_compact_json_snapshot!(
            Request::<Option<()>> {
                jsonrpc: "2.0".to_string(),
                method: "method".to_string(),
                params: None,
                id: 1,
            },
            @r#"{"jsonrpc": "2.0", "method": "method", "id": 1}"#
        );
    }

    #[test]
    fn test_request_deserialization() {
        insta::assert_debug_snapshot!(
            serde_json::from_str::<Request<Vec<i32>>>(r#"{"jsonrpc": "2.0", "method": "method", "params": [42, 23], "id": 1}"#),
            @r###"
        Ok(
            Request {
                jsonrpc: "2.0",
                method: "method",
                params: Some(
                    [
                        42,
                        23,
                    ],
                ),
                id: 1,
            },
        )
        "###
        );
    }

    #[test]
    fn test_notification_serialization() {
        insta::assert_compact_json_snapshot!(
            Notification {
                jsonrpc: "2.0".to_string(),
                method: "method".to_string(),
                params: Some(vec![42, 23]),
            },
            @r###"{"jsonrpc": "2.0", "method": "method", "params": [42, 23]}"###
        );
    }

    #[test]
    fn test_notification_deserialization() {
        insta::assert_debug_snapshot!(
            serde_json::from_str::<Notification<Vec<i32>>>(r#"{"jsonrpc": "2.0", "method": "method", "params": [42, 23]}"#),
            @r###"
        Ok(
            Notification {
                jsonrpc: "2.0",
                method: "method",
                params: Some(
                    [
                        42,
                        23,
                    ],
                ),
            },
        )
        "###
        );
    }

    #[test]
    fn test_response_serialization() {
        insta::assert_compact_json_snapshot!(
            Response {
                jsonrpc: "2.0".to_string(),
                result: Ok(19),
                id: Some(1),
            },
            @r###"{"jsonrpc": "2.0", "result": 19, "id": 1}"###
        );

        insta::assert_compact_json_snapshot!(
            Response::<()> {
                jsonrpc: "2.0".to_string(),
                result: Err(Error {
                    code: -32601,
                    message: "Method not found".to_string(),
                    data: Some(json!(["Some", "data"]))
                }),
                id: None,
            },
            @r###"{"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found", "data": ["Some", "data"]}, "id": null}"###
        );
    }

    #[test]
    fn test_response_deserialization() {
        insta::assert_debug_snapshot!(
            serde_json::from_str::<Response<i32>>(r#"{"jsonrpc": "2.0", "result": 19, "id": 1}"#),
            @r#"
        Ok(
            Response {
                jsonrpc: "2.0",
                result: Ok(
                    19,
                ),
                id: Some(
                    1,
                ),
            },
        )
        "#
        );

        insta::assert_debug_snapshot!(
            serde_json::from_str::<Response<()>>(r#"{"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found", "data": ["Some", "data"]}, "id": null}"#),
            @r#"
        Ok(
            Response {
                jsonrpc: "2.0",
                result: Err(
                    Error {
                        code: -32601,
                        message: "Method not found",
                        data: Some(
                            Array [
                                String("Some"),
                                String("data"),
                            ],
                        ),
                    },
                ),
                id: None,
            },
        )
        "#
        );
    }
}
