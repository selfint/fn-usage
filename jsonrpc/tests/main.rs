use std::sync::mpsc::{Receiver, Sender};

use jsonrpc::{client::Client, types::Request};

macro_rules! jsonrpc_server {
    {
        $msg:expr,
        {
            $(
                --> $a:literal
                <-- $b:literal
            )*
        }
    } => {
        match $msg {
            $($a => $b,)*
            msg => {println!("Got notification: {}", msg); ""}
        }
    }
}

macro_rules! build_request {
    ({"jsonrpc": $jsonrpc:expr, "method": $method:expr, "id": $id:expr}) => {
        Request {
            jsonrpc: $jsonrpc.to_string(),
            method: $method.to_string(),
            params: None,
            id: $id,
        }
    };
    ({"jsonrpc": $jsonrpc:expr, "method": $method:expr, "params": $params:expr, "id": $id:expr}) => {
        Request {
            jsonrpc: $jsonrpc.to_string(),
            method: $method.to_string(),
            params: Some($params),
            id: $id,
        }
    };
}

fn fake_jsonrpc_server(
    client_rx: Receiver<String>,
    server_tx: Sender<String>,
    kill_server_rx: Receiver<()>,
) {
    while kill_server_rx.try_recv().is_err() {
        if let Ok(msg) = client_rx.try_recv() {
            let response = jsonrpc_server! {
                msg.replace(':', ": ").replace(',', ", ").as_ref(),
                {
                    --> r#"{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}"#
                    <-- r#"{"jsonrpc": "2.0", "result": 19, "id": 1}"#
                    --> r#"{"jsonrpc": "2.0", "method": "subtract", "params": [23, 42], "id": 2}"#
                    <-- r#"{"jsonrpc": "2.0", "result": -19, "id": 2}"#
                    --> r#"{"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 3}"#
                    <-- r#"{"jsonrpc": "2.0", "result": 19, "id": 3}"#
                    --> r#"{"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 4}"#
                    <-- r#"{"jsonrpc": "2.0", "result": 19, "id": 4}"#
                    --> r#"{"jsonrpc": "2.0", "method": "foobar", "id": 5}"#
                    <-- r#"{"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found"}, "id": 5}"#
                    --> r#"{"jsonrpc": "2.0", "method": 1, "params": "bar"}"#
                    <-- r#"{"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null}"#
                }
            };

            server_tx
                .send(response.to_string())
                .expect("failed to send response");
        }
    }
}

#[tokio::test]
async fn test_server() {
    let (client_tx, client_rx) = std::sync::mpsc::channel::<String>();
    let (server_tx, server_rx) = std::sync::mpsc::channel::<String>();

    let (kill_server_tx, kill_server_rx) = std::sync::mpsc::channel::<()>();
    let server_handle =
        std::thread::spawn(move || fake_jsonrpc_server(client_rx, server_tx, kill_server_rx));

    let mut client = Client::new(client_tx, server_rx);

    macro_rules! test_request {
        ($params:ty, $result:ty, $error:ty, $($request:tt)*) => {
            insta::assert_debug_snapshot!(
                client
                    .request::<$params, $result, $error>(build_request!($($request)*))
                    .await
            );
        };
        ($result:ty, $error:ty, $($request:tt)*) => {
            insta::assert_debug_snapshot!(
                client
                    .request::<_, $result, $error>(build_request!($($request)*))
                    .await
            );
        };
    }

    #[derive(serde::Serialize, serde::Deserialize, Debug)]
    struct Params {
        subtrahend: i64,
        minuend: i64,
    }

    test_request!(i64, (), {"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1});
    test_request!(i64, (), {"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1});
    test_request!(i64, (), {"jsonrpc": "2.0", "method": "subtract", "params": [23, 42], "id": 2});
    test_request!(i64, (), {"jsonrpc": "2.0", "method": "subtract", "params": Params {subtrahend: 23, minuend: 42}, "id": 3});
    test_request!(i64, (), {"jsonrpc": "2.0", "method": "subtract", "params": Params {minuend: 42, subtrahend: 23}, "id": 4});
    test_request!((), i64, (), {"jsonrpc": "2.0", "method": "foobar", "id": 5});

    kill_server_tx
        .send(())
        .expect("failed to send kill signal to server");
    server_handle.join().expect("failed to join server");
}
