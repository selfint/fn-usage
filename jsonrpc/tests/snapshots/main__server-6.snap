---
source: jsonrpc/tests/main.rs
expression: "client.request::<(), i64,\n        ()>(build_request!({\n                \"jsonrpc\" : \"2.0\", \"method\" : \"foobar\", \"id\" : 5\n            })).await"
---
Ok(
    Response {
        jsonrpc: "2.0",
        result: Error {
            code: -32601,
            message: "Method not found",
            data: None,
        },
        id: Some(
            5,
        ),
    },
)
